pkg/api/unversioned.go

```diff
@@ -22,16 +22,64 @@ import (
 
 // This file contains API types that are unversioned.
 
-// APIVersions lists the api versions that are available, to allow
-// version negotiation. APIVersions isn't just an unnamed array of
-// strings in order to allow for future evolution, though unversioned
+// APIVersions lists the versions that are available, to allow clients to
+// discover the API at /api.
 type APIVersions struct {
+	// Versions are the api versions that are available.
 	Versions []string `json:"versions"`
 }
 
+// APIGroupList is a list of APIGroup, to allow clients to discover the API at
+// /apis.
+type APIGroupList struct {
+	// Groups is a list of APIGroup.
+	Groups []APIGroup `json:"groups"`
+}
+
+// APIGroup contains the name, the supported versions, and the preferred version
+// of a group.
+type APIGroup struct {
+	// GroupName represents the name of the group.
+	GroupName string `json:"groupName"`
+	// GroupVersions are the versions supported in this group.
+	GroupVersions []GroupVersion `json:"groupVersions"`
+	// PreferredVersion is the version preferred by the API server, which
+	// probably is the storage version.
+	PreferredVersion GroupVersion `json: "preferredVersion"`
+}
+
+// GroupVersion contains the "group/version" and "version" string of a version.
+// It is made a struct to keep extensiblity.
+type GroupVersion struct {
+	// GroupVersion is in the form of "group/version".
```

Comment:
We should also mention that this is the form specified in the `apiVersion` field of resources in the APIs.

This made me consider whether this field should also be called `apiVersion`, but I think that would be more confusing. I also considered `versionWithGroup`, but the verbosity doesn't seem worth it and the existing form has the additional advantage of mimicking the form of the values, so `groupVersion` is fine.